File: test files/java/pleno_avancado.java
import java.time.Clock;
import java.util.Optional;

public class PlenoAvancado {
    enum TaskStatus {
        TODO, DOING, DONE, BLOCKED
    }

    record Task(int id, String title, Integer ownerId, TaskStatus status, int estimateHours) {}

    record CompleteTaskOutput(int taskId, String completedAt) {}

    interface TaskRepository {
        Optional<Task> findById(int id);
        Task save(Task task);
    }

    interface AuditGateway {
        void publish(int taskId, String action, String at);
    }

    static class TaskPolicy {
        Optional<String> validate(Task task) {
            if (task.ownerId() == null) {
                return Optional.of("TASK_WITHOUT_OWNER");
            }

            if (task.status() == TaskStatus.BLOCKED) {
                return Optional.of("TASK_BLOCKED");
            }

            if (task.estimateHours() <= 0) {
                return Optional.of("INVALID_ESTIMATE");
            }

            return Optional.empty();
        }
    }

    static class CompleteTaskUseCase {
        private final TaskRepository taskRepository;
        private final AuditGateway auditGateway;
        private final Clock clock;
        private final TaskPolicy taskPolicy;

        CompleteTaskUseCase(TaskRepository taskRepository, AuditGateway auditGateway, Clock clock, TaskPolicy taskPolicy) {
            this.taskRepository = taskRepository;
            this.auditGateway = auditGateway;
            this.clock = clock;
            this.taskPolicy = taskPolicy;
        }

        Optional<CompleteTaskOutput> execute(int taskId) {
            Optional<Task> maybeTask = taskRepository.findById(taskId);
            if (maybeTask.isEmpty()) {
                return Optional.empty();
            }

            Task task = maybeTask.get();
            if (taskPolicy.validate(task).isPresent()) {
                return Optional.empty();
            }

            String completedAt = clock.instant().toString();
            Task updatedTask = new Task(task.id(), task.title(), task.ownerId(), TaskStatus.DONE, task.estimateHours());
            taskRepository.save(updatedTask);
            auditGateway.publish(updatedTask.id(), "task.completed", completedAt);
            return Optional.of(new CompleteTaskOutput(updatedTask.id(), completedAt));
        }
    }
}

File: test files/typescript/pleno_avancado.ts
type TaskStatus = 'todo' | 'doing' | 'done' | 'blocked';

type DomainError =
  | 'TASK_NOT_FOUND'
  | 'TASK_WITHOUT_OWNER'
  | 'TASK_BLOCKED'
  | 'INVALID_ESTIMATE';

type Result<T> =
  | { ok: true; data: T }
  | { ok: false; error: DomainError };

interface Task {
  id: number;
  title: string;
  ownerId: number | null;
  status: TaskStatus;
  estimateHours: number;
}

interface TaskRepository {
  findById(id: number): Promise<Task | null>;
  save(task: Task): Promise<Task>;
}

interface AuditGateway {
  publish(entry: { taskId: number; action: string; at: string }): Promise<void>;
}

interface Clock {
  now(): Date;
}

class TaskCompletionPolicy {
  validate(task: Task): Result<Task> {
    if (task.ownerId === null) {
      return { ok: false, error: 'TASK_WITHOUT_OWNER' };
    }

    if (task.status === 'blocked') {
      return { ok: false, error: 'TASK_BLOCKED' };
    }

    if (task.estimateHours <= 0) {
      return { ok: false, error: 'INVALID_ESTIMATE' };
    }

    return { ok: true, data: task };
  }
}

class CompleteTaskUseCase {
  constructor(
    private readonly taskRepository: TaskRepository,
    private readonly auditGateway: AuditGateway,
    private readonly clock: Clock,
    private readonly policy: TaskCompletionPolicy
  ) {}

  async execute(taskId: number): Promise<Result<{ taskId: number; completedAt: string }>> {
    const task = await this.taskRepository.findById(taskId);
    if (!task) {
      return { ok: false, error: 'TASK_NOT_FOUND' };
    }

    const validation = this.policy.validate(task);
    if (!validation.ok) {
      return validation;
    }

    const completedAt = this.clock.now().toISOString();
    const updatedTask = { ...task, status: 'done' as const };

    await this.taskRepository.save(updatedTask);
    await this.auditGateway.publish({
      taskId: updatedTask.id,
      action: 'task.completed',
      at: completedAt,
    });

    return {
      ok: true,
      data: {
        taskId: updatedTask.id,
        completedAt,
      },
    };
  }
}
